x/ccv/provider/keeper/validator_set_update.go

```diff
@@ -174,11 +128,53 @@ func DiffValidators(
 	return updates
 }
 
-// SetConsumerValSet resets the current consumer validators with the `nextValidators` computed by
-// `ComputeNextEpochConsumerValSet` and hence this method should only be called after `ComputeNextEpochConsumerValSet` has completed.
-func (k Keeper) SetConsumerValSet(ctx sdk.Context, chainID string, nextValidators []types.ConsumerValidator) {
-	k.DeleteConsumerValSet(ctx, chainID)
-	for _, val := range nextValidators {
-		k.SetConsumerValidator(ctx, chainID, val)
+// CreateConsumerValidator creates a consumer validator for `chainID` from the given staking `validator`
+func (k Keeper) CreateConsumerValidator(ctx sdk.Context, chainID string, validator stakingtypes.Validator) (types.ConsumerValidator, error) {
+	power := k.stakingKeeper.GetLastValidatorPower(ctx, validator.GetOperator())
+	consAddr, err := validator.GetConsAddr()
+	if err != nil {
+		return types.ConsumerValidator{}, fmt.Errorf("could not retrieve validator's (%+v) consensus address: %w",
+			validator, err)
+	}
+
+	consumerPublicKey, foundConsumerPublicKey := k.GetValidatorConsumerPubKey(ctx, chainID, types.NewProviderConsAddress(consAddr))
+	if !foundConsumerPublicKey {
+		consumerPublicKey, err = validator.TmConsPublicKey()
+		if err != nil {
+			return types.ConsumerValidator{}, fmt.Errorf("could not retrieve validator's (%+v) public key: %w", validator, err)
+		}
+	}
+
+	return types.ConsumerValidator{
+		ProviderConsAddr:  consAddr,
+		Power:             power,
+		ConsumerPublicKey: &consumerPublicKey,
+	}, nil
+}
+
+// ComputeNextEpochConsumerValSet returns the next validator set that is responsible for validating consumer
+// chain `chainID`. The next validator set corresponds to the `bondedValidator`s that `shouldConsider` evaluates to `true`.
+func (k Keeper) ComputeNextEpochConsumerValSet(
+	ctx sdk.Context,
+	chainID string,
+	bondedValidators []stakingtypes.Validator,
+	shouldConsider func(validator stakingtypes.Validator) bool,
```

Comment:
Is ShouldConsider the thing we would use for what Jehan calls power shaping?
Because I think a predicate (val -> bool) is probably not expressive enough.
In reality, we want to be able to put functions
[]ConsumerValidator -> []ConsumerValidator

The predicate is not enough to express e.g. power caps

Reply:
We discussed offline: We agree that this can't be used for power shaping. But it is something we can do add/refactor later when we actually implement power shaping, etc.